Add Fedora Root On ZFS guide and in general revamp Fedora's wiki (#21)
openzfs/openzfs-docs: docs/Getting Started/Fedora/Fedora Root On ZFS.rst

        4 KiB logical sectors (“4Kn” drives) to meet the minimum cluster size
        (given the partition size of 512 MiB) for FAT32. It also works fine on
        drives which present 512 B sectors.

#. Set a root password::

     passwd

#. Optional (but recommended): Mount a tmpfs to ``/tmp``

   If you chose to create a ``/tmp`` dataset above, skip this step, as they
   are mutually exclusive choices. Otherwise, you can put ``/tmp`` on a
   tmpfs (RAM filesystem) by enabling the ``tmp.mount`` unit.

   ::

     cp /usr/share/systemd/tmp.mount /etc/systemd/system/
     systemctl enable tmp.mount

.. note::
   GRUB installation is not supported by this guide and will not be supported in the forseeable future. The above steps should have installed systemd-boot, an alternative to GRUB which provides the majority of GRUBS features. If you still wish to use GRUB, it might be possible to chainload systemd-boot using GRUB and boot your Fedora installation that way. Instructions on how to do this will not be provided and this has not been tested to work. You are welcome to make a PR for this however

Step 6: First Boot

> ------------------

Reviewer: Before the boot is it need zfs-list.cache files for rpool and bpool.

Reviewer: I tried it without zfs-list.cache files for pools it was crashed...

     ``atime`` behavior (with its significant performance impact) and
     ``atime=off`` (which provides the best performance by completely
     disabling atime updates). Since Linux 2.6.30, ``relatime`` has been
     the default for other filesystems. See `RedHat’s documentation
     <https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/power_management_guide/relatime>`__
     for further information.
   - Setting ``xattr=sa`` `vastly improves the performance of extended
     attributes
     <https://github.com/zfsonlinux/zfs/commit/82a37189aac955c81a59a5ecc3400475adb56355>`__.
     Inside ZFS, extended attributes are used to implement POSIX ACLs.
     Extended attributes can also be used by user-space applications.
     `They are used by some desktop GUI applications.
     <https://en.wikipedia.org/wiki/Extended_file_attributes#Linux>`__
     `They can be used by Samba to store Windows ACLs and DOS attributes;
     they are required for a Samba Active Directory domain controller.
     <https://wiki.samba.org/index.php/Setting_up_a_Share_Using_Windows_ACLs>`__
     Note that ``xattr=sa`` is `Linux-specific

>      <http://open-zfs.org/wiki/Platform_code_differences>`__. If you move your

Reviewer 2: ```suggestion
     <https://openzfs.org/wiki/Platform_code_differences>`__. If you move your
```

   ``canmount=noauto``.

#. Create datasets::

     zfs create                                 rpool/home
     zfs create -o mountpoint=/root             rpool/home/root
     zfs create -o canmount=off                 rpool/var
     zfs create -o canmount=off                 rpool/var/lib
     zfs create                                 rpool/var/log
     zfs create                                 rpool/var/spool

   The datasets below are optional, depending on your preferences and/or
   software choices.

   If you wish to exclude these from snapshots::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/cache
     zfs create -o com.sun:auto-snapshot=false  rpool/var/tmp
     chmod 1777 /mnt/var/tmp

   If this system will use Docker (which manages its own datasets &
   snapshots)::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/lib/docker

   If this system will use NFS (locking)::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/lib/nfs

   A tmpfs is recommended later, but if you want a separate dataset for
   ``/tmp``::

     zfs create -o com.sun:auto-snapshot=false  rpool/tmp
     chmod 1777 /mnt/tmp
   

>    Note that the reason why we are not fully seperating everything like we did in Ubuntu is because dnf will fail to install or update certain packages if we create too many datasets. An example of one such package is filesystem, which fails to install if other ZFS datasets are created.

Reviewer 2: ```suggestion
   Note that the reason why we are not fully separating everything like we did in Ubuntu is because dnf will fail to install or update certain packages if we create too many datasets. An example of one such package is filesystem, which fails to install if other ZFS datasets are created.
```

   The primary goal of this dataset layout is to separate the OS from user
   data. This allows the root filesystem to be rolled back without rolling
   back user data.

   If you do nothing extra, ``/tmp`` will be stored as part of the root
   filesystem. Alternatively, you can create a separate dataset for ``/tmp``,
   as shown above. This keeps the ``/tmp`` data out of snapshots of your root
   filesystem. It also allows you to set a quota on ``rpool/tmp``, if you want
   to limit the maximum space used. Otherwise, you can use a tmpfs (RAM
   filesystem) later.

#. Copy the LiveCD to your HDD/SDD::

     rsync -avxHASX / /mnt/
   
   It is important to not forget the trailing /.

>    This command copies the LiveCD to our new zfs datasets and this is the only way I have found to reliably install and boot Fedora Workstation

Reviewer 2: ```suggestion
   This command copies the LiveCD to our new zfs datasets and this is the only way guide's author found to reliably install and boot Fedora Workstation
```

.. highlight:: sh

Fedora Root on ZFS
==================

.. contents:: Table of Contents
  :local:

Overview
--------

Caution
~~~~~~~

- This HOWTO uses a whole physical disk.
- Do not use these instructions for dual-booting.
- Backup your data. Any existing data will be lost.

System Requirements
~~~~~~~~~~~~~~~~~~~

> - `64-bit Fedora Workstation (spins are not yet tested yet)

Reviewer 3: second "yet" unnecessary

  <https://getfedora.org/en/workstation/download/>`__
- `A 64-bit kernel is strongly encouraged.
  <https://openzfs.github.io/openzfs-docs/Project%20and%20Community/FAQ.html#bit-vs-64-bit-systems>`__
- Installing on a drive which presents 4 KiB logical sectors (a “4Kn” drive)
  only works with UEFI booting. This not unique to ZFS. `GRUB does not and
  will not work on 4Kn with legacy (BIOS) booting.
  <http://savannah.gnu.org/bugs/?46700>`__

> - You MUST use a UEFI system for this guide as GRUB is not supported and will not be supported in the forseeable future. To verify that you are using UEFI, go to /sys/firmware/efi/efivars and make sure it is not empty.

Reviewer 3: It's worth to note that Fedora developers are discussing dropping legacy BIOS support entirely in future releases: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/QBANCA2UAJ5ZSMDVVARLIYAJE66TYTCD/

Encryption
~~~~~~~~~~

This guide supports three different encryption options: unencrypted, ZFS
native encryption, and LUKS. With any option, all ZFS features are fully
available.

Unencrypted does not encrypt anything, of course. With no encryption
happening, this option naturally has the best performance.

ZFS native encryption encrypts the data and most metadata in the root
pool. It does not encrypt dataset or snapshot names or properties. The
boot pool is not encrypted at all, but it only contains the bootloader,
kernel, and initrd. (Unless you put a password in ``/etc/fstab``, the
initrd is unlikely to contain sensitive data.) The system cannot boot
without the passphrase being entered at the console. Performance is
good. As the encryption happens in ZFS, even if multiple disks (mirror
or raidz topologies) are used, the data only has to be encrypted once.

LUKS encrypts almost everything. The only unencrypted data is the bootloader,
kernel, and initrd. The system cannot boot without the passphrase being
entered at the console. Performance is good, but LUKS sits underneath ZFS, so
if multiple disks (mirror or raidz topologies) are used, the data has to be

> encrypted once per disk. Note that I (cheesycod) do not use LUKS and that you are using this at your own risk.

Reviewer 2: This is an official project documentation, let's not use ` I ` statements. And, if we are not sure on anything - we may not want to write about it.

Fox ex, you can write here:
```suggestion
encrypted once per disk. 

.. note::
This manual describes LUKS configuration only as an example, it wasn't tested on everyday basis usage.
```

#. Partition your disk(s):

   Run this to create your ESP::

     sgdisk     -n1:1M:+512M   -t1:EF00 $DISK

   (Optional, but recommended if you have high memory pressure): Create a swap partition::

     sgdisk     -n2:0:+<size>G  -t2:8200 $DISK # Make sure you replace <size> with the size of your swap partition.
     mkswap     $DISK-part2
     swapon     $DISK-part2

   .. note::
      Creation of the swap partition shoud not be done if you plan on using ZFS for swap. Instead, please follow Step 8 instead

   Choose one of the following options:

   - Unencrypted or ZFS native encryption (change the -n3 and -t3 to -n2 and -t3 respectively if you added a swap partition)::

       sgdisk     -n3:0:0        -t3:BF00 $DISK

   - LUKS (same warning as with Unencrypted and ZFS native encryption, change the -n3 and -t3 to -n2 and -t3 if you are not adding swap)::

       sgdisk     -n3:0:0        -t3:8309 $DISK

   If you are creating a mirror or raidz topology, repeat the partitioning
   commands for all the disks which will be part of the pool.

> #. Create the root pool (change the disk to ${DISK}-part2 if you did):

Reviewer 3: if you did what?

   Choose one of the following options:

   - Unencrypted::

       zpool create \
           -o ashift=12 \
           -O acltype=posixacl -O canmount=off -O compression=lz4 \
           -O dnodesize=auto -O normalization=formD -O relatime=on \
           -O xattr=sa -O mountpoint=/ -R /mnt \
           rpool ${DISK}-part3 # ${DISK}-part2 if swap partition was not created

   - ZFS native encryption::

       zpool create \
           -o ashift=12 \
           -O encryption=aes-256-gcm \
           -O keylocation=prompt -O keyformat=passphrase \
           -O acltype=posixacl -O canmount=off -O compression=lz4 \
           -O dnodesize=auto -O normalization=formD -O relatime=on \
           -O xattr=sa -O mountpoint=/ -R /mnt \
           rpool ${DISK}-part2 # ${DISK}-part2 if swap partition was not created

   - LUKS::

>        apt install --yes cryptsetup

Reviewer 3: `apt`?

   ``canmount=noauto``.

#. Create datasets::

     zfs create                                 rpool/home
     zfs create -o mountpoint=/root             rpool/home/root
     zfs create -o canmount=off                 rpool/var
     zfs create -o canmount=off                 rpool/var/lib
     zfs create                                 rpool/var/log
     zfs create                                 rpool/var/spool

   The datasets below are optional, depending on your preferences and/or
   software choices.

   If you wish to exclude these from snapshots::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/cache
     zfs create -o com.sun:auto-snapshot=false  rpool/var/tmp
     chmod 1777 /mnt/var/tmp

   If this system will use Docker (which manages its own datasets &
   snapshots)::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/lib/docker

   If this system will use NFS (locking)::

     zfs create -o com.sun:auto-snapshot=false  rpool/var/lib/nfs

   A tmpfs is recommended later, but if you want a separate dataset for
   ``/tmp``::

     zfs create -o com.sun:auto-snapshot=false  rpool/tmp
     chmod 1777 /mnt/tmp
   

>    Note that the reason why we are not fully seperating everything like we did in Ubuntu is because dnf will fail to install or update certain packages if we create too many datasets.

Reviewer 3: can you add a comment with an example of such package

Author: Yes, basesystem/filesystem (I forgot which one, need to check) will fail if those datasets are present.

   The primary goal of this dataset layout is to separate the OS from user
   data. This allows the root filesystem to be rolled back without rolling
   back user data.

   If you do nothing extra, ``/tmp`` will be stored as part of the root
   filesystem. Alternatively, you can create a separate dataset for ``/tmp``,
   as shown above. This keeps the ``/tmp`` data out of snapshots of your root
   filesystem. It also allows you to set a quota on ``rpool/tmp``, if you want
   to limit the maximum space used. Otherwise, you can use a tmpfs (RAM
   filesystem) later.

#. Copy the LiveCD to your HDD/SDD::

     rsync -avxHASX / /mnt/
   
   It is important to not forget the trailing /.

>    This command copies the LiveCD to our new zfs datasets and this is the only way I have found to reliably install and boot Fedora Workstation

Reviewer 3: what issues did you have during the installation of the `Fedora Workstation` packages group?

Author: System fails to boot and hangs at Starting GNOME Desktop Manager

Author: TTY also fails at this point and doesn't allow you to login

Step 4: System Configuration
----------------------------

#. Configure the hostname:

   Replace ``HOSTNAME`` with the desired hostname::

     echo HOSTNAME > /mnt/etc/hostname
     vi /mnt/etc/hosts

   .. code-block:: text

     Add a line:
     127.0.1.1       HOSTNAME
     or if the system has a real name in DNS:
     127.0.1.1       FQDN HOSTNAME

   **Hint:** Use ``nano`` or ``vim`` if you find ``vi`` confusing.

#. Configure the network interface:

   Find the interface name::

     ip addr show

   Adjust ``NAME`` below to match your interface name::

>      vi /mnt/etc/network/interfaces.d/NAME

Reviewer 3: it is 2020, it's time to stop using old config files and either use `nmconnection` in `/etc/NetworkManager/system-connections/` when using NetworkManager which is default or `network` unit files in `/etc/systemd/network` when using `systemd-networkd`

Reviewer 4: This isn't so much an issue of _old_ config files as this is a Debian-ism that hasn't been ported to Fedora's way yet.

Troubleshooting
---------------

Rescuing using a Live CD
~~~~~~~~~~~~~~~~~~~~~~~~

Go through `Step 1: Prepare The Install Environment
<#step-1-prepare-the-install-environment>`__.

For LUKS, first unlock the disk(s)::

  dnf install cryptsetup
  cryptsetup luksOpen /dev/disk/by-id/scsi-SATA_disk1-part4 luks1
  # Repeat for additional disks, if this is a mirror or raidz topology.

Mount everything correctly::

  zpool export -a
  zpool import -N -R /mnt rpool
  zfs load-key -a
  zfs mount rpool/ROOT/fedora
  zfs mount -a

If needed, you can chroot into your installed environment::

  mount --rbind /dev  /mnt/dev
  mount --rbind /proc /mnt/proc
  mount --rbind /sys  /mnt/sys
  chroot /mnt /bin/bash --login
  mount /boot
  mount -a

Do whatever you need to do to fix your system.

When done, cleanup::

  exit
  mount | grep -v zfs | tac | awk '/\/mnt/ {print $3}' | \
      xargs -i{} umount -lf {}
  zpool export -a
  reboot

QEMU/KVM/XEN
~~~~~~~~~~~~

Set a unique serial number on each virtual disk using libvirt or qemu
(e.g. ``-drive if=none,id=disk1,file=disk1.qcow2,serial=1234567890``).

To be able to use UEFI in guests (instead of only BIOS booting), run
this on the host::

>   sudo apt install ovmf # or dnf install edk2-ovmf (if the host is Fedora)

Reviewer 3: `apt`?

- `A 64-bit kernel is strongly encouraged.
  <https://openzfs.github.io/openzfs-docs/Project%20and%20Community/FAQ.html#bit-vs-64-bit-systems>`__
- Installing on a drive which presents 4 KiB logical sectors (a “4Kn” drive)
  only works with UEFI booting. This not unique to ZFS. `GRUB does not and
  will not work on 4Kn with legacy (BIOS) booting.
  <http://savannah.gnu.org/bugs/?46700>`__
- You MUST use a UEFI system for this guide as GRUB is not yet supported. To verify UEFI, go to /sys/firmware/efi/efivars and make sure it is not empty.

Computers that have less than 2 GiB of memory run ZFS slowly. 4 GiB of memory
is recommended for normal performance in basic workloads. If you wish to use
deduplication, you will need `massive amounts of RAM
<http://wiki.freebsd.org/ZFSTuningGuide#Deduplication>`__. Enabling
deduplication is a permanent change that cannot be easily reverted.

Support
~~~~~~~

If you need help, reach out to the community using the :doc:`zfs-discuss
mailing list <../../Project and Community/Mailing Lists>` or IRC at
`#zfsonlinux <irc://irc.freenode.net/#zfsonlinux>`__ on `freenode
<https://freenode.net/>`__. If you have a bug report or feature request

> related to this HOWTO, please `file a new issue and mention @rlaager

Reviewer 4: I don't use Fedora. Take my name out. If you're going to maintain this, you can sub in your username. Also, update the URL below for filing an issue, which has my username in it and also has the distro name and version (so if someone clicks the link and keeps the suggested text, I know which HOWTO we are talking about).

Author: Ok, fixed

<https://github.com/openzfs/openzfs-docs/issues/new?body=@rlaager,%20I%20have%20the%20following%20issue%20with%20the%20Debian%20Buster%20Root%20on%20ZFS%20HOWTO:>`__.

Contributing
~~~~~~~~~~~~

#. Fork and clone: https://github.com/openzfs/openzfs-docs

#. Install the tools::

    sudo dnf install python3-pip sensible-utils
    cd openzfs-docs/docs
    pip3 install -r requirements.txt
    # Add ~/.local/bin to your $PATH, e.g. by adding this to ~/.bashrc:
    PATH=$HOME/.local/bin:$PATH

#. Make your changes.

#. Test::

    cd docs # If you aren't already in the docs folder
    make html
    sensible-browser _build/html/index.html

#. ``git commit --signoff`` to a branch, ``git push``, and create a pull

>    request. Mention @rlaager.

Reviewer 4: Likewise I'm mentioned here.

Author: Yeah, I was just focusing on writing the guide. I'll fix this now

Reviewer 4: No worries. I understand it's a draft.

Author: It should be fixed now.

     sed -i 's/#timeout.*/timeout 10/' /boot/loader/loader.conf

#. Ensure that zfs-target is enabled::

>      systemctl enable zfs.target

Reviewer 5: I've recently encountered a Fedora system that failed to mount the zfs pools during boot because the zfs kernel module was not available in time. I found that adding `rd.driver.pre=zfs` to the list of kernel parameters resolved the problem. I suspect that may be what is needed here instead of `systemctl enable zfs.target`. The zfs target should be enabled by default.

Just FYI.

Author: the zfs.target isnt to solve thats, its to solve the edge case in which the zfs pools and partitions arent loaded for some reason

Reviewer 5: If there is even a chance that the system could fail to boot due to sub-filesystems failing to mount, I'd be tempted to minimize the layout down to just / and /home (if /home fails to mount, it will not prevent the system from booting and it is easy to mount after booting using the root account). It looks like your guide is pretty close to doing that already. Also, I _think_ that is pretty close to how Fedora 33 is going to layout their default btrfs file system: https://fedoramagazine.org/btrfs-coming-to-fedora-33/#comment-502491

I see that your current version has /var/log and /var/spool as separate file systems. If they were to fail to mount, the system should still boot, but it might be good to note that if they fail once, they will continue to fail on subsequent boots/attempts if any files are left under those mountpoints. So it might be helpful to inform the user that they will need to delete auto-generated files under those directories to recover from the error. Or better yet, IMHO, just keep them on the / file system by default and let the user know that if they don't want their logs rolled back, then they will need to make /var/log a separate file system.

I still haven't tried it out, but your guide is looking prettying good so far. Thanks for doing the research on this.

Author: Your welcome
